### ISSUE TITLE
VmWare ESXi support

### ISSUE THREAD

**Issue by chris-rock** [inspec/inspec#3133]
We want to bring full support for VmWare into InSpec. VmWare is considered similar to AWS and Azure as a separate platform. Therefore we want to be able to connect to VmWare ESXi host via `inspec exec test.rb -t vmware://user:password@host`

At this point, we have two competing solutions with their own advantages and disadvantages

* [rbvmomi](https://github.com/vmware/rbvmomi )
* [PowerCLI](https://www.vmware.com/support/developer/PowerCLI/ )

**rbvmomi**

We've written the [inspec-vmware](https://github.com/chef/inspec-vmware ) that loads the [rbvmomi](https://github.com/vmware/rbvmomi ) gems. We've implemented a vmware-hardening-guide based on that to test the implementation. 

Pro:
- technically it works similar to existing AWS and Azure solutions by using a ruby gem

Con:
- no long-term support for rbvmomi is questionable, we already see that not all features are available
- difficult to understand and extend for existing VmWare administrators, who use PowerCLI

**PowerCLI**

@jjasghar implemented an InSpec profile prototype based on VmWare PowerCLI. This approach requires Powershell with the PowerCLI extensions. Recent improvements of Powershell and PowerCLI allow us to run this implementation in Powershell under Linux. This would allow us to ship it with InSpec and Automate for the supported platforms.

The challenging part is that we run a InSpec test against a local target that tests a remote target with `inspec exec vmware-profile`. Then you use attributes in the tests to connect to the remote target.

```
# no need for
viserver = attribute('viserver', default: 'vcenter.example.com', description: 'The server you want to connect to')
username = attribute('username', default: 'root', description: 'Username to connect as')
password = attribute('password', default: 'Cod3Can!', description: 'Password to connect with')

ad_local_user_script = <<~EOH
  Connect-VIserver "#{viserver}" -User "#{username}" -Password "#{password}" | Out-Null
  Get-VMhost | Get-VMHostAuthentication | select DomainMembershipStatus | ConvertTo-Json
EOH

ad_local_user_script_data = powershell(ad_local_user_script).stdout

```

This is counter-intuitive for InSpec users, since it conflicts with the existing AWS and Azure platforms.

Pro:
- uses well-supported PowerCLI
- Admins know that tool very well

Con 
- counter-intuitive target

## Solution

As I discussed with @arlimus , he proposed a solution where are able to combine the best of both worlds.

TLDR:
- we use Powershell
- we use `inspec exec test.rb -t vmware://user:password@host`
- we abstract the PowerCLI connection
- all vmware scripts go via a new `powercli` resource

The end-users will test against vmware as a target with 

```
inspec exec test.rb -t vmware://user:password@host
```

Now, instead of 

```
# no need for
viserver = attribute('viserver', default: 'vcenter.example.com', description: 'The server you want to connect to')
username = attribute('username', default: 'root', description: 'Username to connect as')
password = attribute('password', default: 'Cod3Can!', description: 'Password to connect with')

ad_local_user_script = <<~EOH
  Connect-VIserver "#{viserver}" -User "#{username}" -Password "#{password}" | Out-Null
  Get-VMhost | Get-VMHostAuthentication | select DomainMembershipStatus | ConvertTo-Json
EOH

ad_local_user_script_data = powershell(ad_local_user_script).stdout
```

we are going to abstract the attributes and connection part via train. This would allow us to keep an open powershell session to vmware, similar to our pipeline approach that we use for windows local execution. Train would need to establish the session with:

```
# abstract connect and keep powershell session open
Connect-VIserver "#{viserver}" -User "#{username}" -Password "#{password}" | Out-Null
```

As a user, I can just write my powershell scripts directly:

```
ad_local_user_script = <<~EOH
  Get-VMhost | Get-VMHostAuthentication | select DomainMembershipStatus | ConvertTo-Json
EOH

ad_local_user_script_data = powercli(ad_local_user_script).stdout
```

## Open questions

Open Questions:
- Clarify legal situation, are we able to ship PowerCLI?
- How do we update powercli scripts?
- Can we keep an open session in powershell with a remote connection?
- How do we do use a long-running powershell session?

**Comment by jjasghar** [comment]
https://github.com/jjasghar/powercli-inspec is a custom resource to help wrap PowerCLI for a stop gap.

**Comment by jerryaldrichiii** [comment]
Alright friends, I think we are close to closing this issue. There is now a transport in Train and options in InSpec to make it work. I think after https://github.com/chef/inspec-vmware/pull/23 is merged and we move `inspec-vmware` under the `inspec` GitHub organization we will be ready to close.

Thoughts?

**Comment by chris-rock** [comment]
Thank you all for this great work. The resource pack is available now under https://github.com/inspec/inspec-vmware
